code/modules/library/admin_only.dm

```diff
@@ -0,0 +1,354 @@
+			AND title LIKE CONCAT('%',:title,'%')
+			[book_id ? "AND id LIKE CONCAT('%', :book_id, '%')" : ""]
+			AND (:category = 'Any' OR category = :category)
+			AND ckey LIKE CONCAT('%',:ckey,'%')
+	"}, list("author" = author, "title" = title, "book_id" = book_id, "category" = category, "ckey" = ckey))
+
+	if(!query_library_count_books.warn_execute())
+		qdel(query_library_count_books)
+		return
+	if(query_library_count_books.NextRow())
+		bookcount = text2num(query_library_count_books.item[1])
+	qdel(query_library_count_books)
+
+	page_count = round(max(bookcount - 1, 0) / BOOKS_PER_PAGE) //This is just floor()
+	search_page = clamp(search_page, 0, page_count)
+
+/obj/machinery/computer/libraryconsole/admin_only_do_not_map_in_you_fucker/ui_status(mob/user)
+	if(!check_rights_for(user.client, R_BAN))
+		return UI_CLOSE
+	if(!SSdbcore.Connect())
+		can_connect = FALSE
+		return UI_CLOSE
+	return UI_INTERACTIVE
+
+/obj/machinery/computer/libraryconsole/admin_only_do_not_map_in_you_fucker/ui_act(action, params, datum/tgui/ui)
+	. = ..()
+	if(.)
+		// We'll always trigger a search attempt if the parent does something, this ensures the ui is v fast to update
+		INVOKE_ASYNC(src, PROC_REF(update_db_info))
+		return
+	switch(action)
+		if("set_search_ckey")
+			ckey = params["ckey"]
+			INVOKE_ASYNC(src, PROC_REF(update_db_info))
+			return TRUE
+		if("refresh")
+			last_search_hash = ""
+			INVOKE_ASYNC(src, PROC_REF(update_db_info))
+			return TRUE
+		if("hide_book")
+			var/reason = params["delete_reason"]
+			var/id = params["book_id"]
+			var/client/actor = ui.user?.client
+			if(!actor)
+				return
+			INVOKE_ASYNC(src, PROC_REF(hide_book), id, reason, actor)
+			return TRUE
+		if("unhide_book")
+			var/reason = params["free_reason"]
+			var/id = params["book_id"]
+			var/client/actor = ui.user?.client
+			if(!actor)
+				return
+			INVOKE_ASYNC(src, PROC_REF(unhide_book), id, reason, actor)
+			return TRUE
+		if("get_history")
+			var/id = params["book_id"]
+			book_history["[id]"] = get_book_history(id)
+			return TRUE
+		if("view_book")
+			var/id = params["book_id"]
+			view_book(id, ui.user)
+			return TRUE
+		if("toggle_raw")
+			view_raw = !view_raw
+			return TRUE
+		if("toggle_deleted")
+			show_deleted = !show_deleted
+			INVOKE_ASYNC(src, PROC_REF(update_db_info))
+			return TRUE
+
+/obj/machinery/computer/libraryconsole/admin_only_do_not_map_in_you_fucker/ui_data(mob/user)
+	. = ..()
+	.["view_raw"] = view_raw
+	.["show_deleted"] = show_deleted
+	var/list/histories = list()
+	for(var/id as anything in book_history)
+		var/list/insert = list()
+		for(var/datum/book_history_entry/entry in book_history[id])
+			insert += list(entry.serialize())
+		histories[id] = insert
+	.["history"] = histories
+
+/obj/machinery/computer/libraryconsole/admin_only_do_not_map_in_you_fucker/proc/view_book(id, mob/show_to)
+	if (!SSdbcore.Connect())
+		can_connect = FALSE
+		message_admins("Could not connect to the DB")
```

Comment:
```suggestion
		message_admins("Failed to establish database connection.")
```

same message used everywhere else, though typically it's in a `to_chat(show_to, ...)` kinda affair.